### PR TITLE
Initial PR from dev branch with modules for short-read assembly and mapping

### PR DESCRIPTION
**Pull request by elizabethmcd**
This pull request adds modules, subworkflows, and a workflow for QC, assembly, and mapping of short metagenomic reads. 

Nf-core Modules: 
- fastp 
- spades
- metabat2/jgisummarizecontigs 

Local Modules (the nf-core modules of these don't work for downstream purposes): 
- Bowtie2 alignment 
- Bowtie2 index building

Subworkflows: 
- mapping_depth.nf implements bowtie2 index building and alignment, samtools, and metabat2/jgisummarizecontigs to map to the assembly and get a depth table 

This is implemented in the metagenomic-sr ("short read") workflow.

### CODE REVIEW

**Comment by taylorreiter** [comment]
Do you want me to review this even though tests are failing?

+ The test ignored all look fine.
+ The test warnings all point to TODO strings. Are you ok with these in the state that they are for this PR, or do you want to address some of them now? (i know things like zenodo doi aren't ready to be addressed)
+ I don't know how to handle the failed tests. Are you ok with these failing, or should those things be fixed?

**Comment by elizabethmcd** [comment]
- For the purposes of this PR, I think it's fine to have some TODO's left over. 
- I'm also not sure how to deal with failed tests particularly with some of the linting, I think for now if we merge and then deal with what linting to keep/remove? 
- The test data failed because I haven't specified test data yet, that will be a separate PR

File: CITATIONS.md [hunk 1]
```diff
@@ -11,9 +11,18 @@
...
+- [Fastp](https://doi.org/10.1093/bioinformatics/bty560)
+  > Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu; fastp: an ultra-fast all-in-one FASTQ preprocessor, Bioinformatics, Volume 34, Issue 17, 1 September 2018, Pages i884–i890
+- [SPAdes](https://currentprotocols.onlinelibrary.wiley.com/doi/abs/10.1002/cpbi.102)
+  > Andrey Prjibelski,Dmitry Antipov,Dmitry Meleshko,Alla Lapidus,Anton Korobeynikov. Using SPAdes De Novo Assembler. Current Protocols in Bioinformatics. 2020.
+- [Bowtie2](https://www.nature.com/articles/nmeth.1923)
+  > Ben Langmead and Steven L Salzberg. Fast-gapped read alignment with bowtie2. Nature Methods. 2012.
+- [metabat2 JGI Summarize Contigs](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6662567/)
+  > Dongwan D. Kang,1 Feng Li,2 Edward Kirton,1 Ashleigh Thomas,1 Rob Egan,1 Hong An,2 and Zhong Wang. MetaBAT 2: an adaptive binning algorithm for robust and efficient genome reconstruction from metagenome assemblies. PeerJ. 2019.
+- [Samtools](https://academic.oup.com/gigascience/article/10/2/giab008/6137722?login=false)
```
Comment (taylorreiter) [inline CITATIONS.md:24]: is this the best citation for samtools?

File: modules/local/bowtie2_assembly_align.nf [hunk 2]
```diff
@@ -0,0 +1,44 @@
+//https://github.dev/nf-core/mag
```
Comment (taylorreiter) [inline modules/local/bowtie2_assembly_align.nf:1]: oh great doc, thanks for adding this

File: modules/local/bowtie2_assembly_build.nf [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    tuple val(meta), path(assembly)
+
+    output:
+    tuple val(meta), path(assembly), path('bt2_index_base*'), emit: assembly_index
+    path "versions.yml"                                     , emit: versions
+
+    script:
+    def args = task.ext.args ?: ''
+    """
+    mkdir bowtie
```
Comment (taylorreiter) [inline modules/local/bowtie2_assembly_build.nf:21]: I'm assuming this is fine since it came from nf-core/mag but this feels weird to me. it feels irrelevant? where is it used?

File: modules/nf-core/fastp/main.nf [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        'quay.io/biocontainers/fastp:0.23.2--h79da9fb_0' }"
+
+    input:
+    tuple val(meta), path(reads)
+    val   save_trimmed_fail
+    val   save_merged
+
+    output:
+    tuple val(meta), path('*.fastp.fastq.gz') , optional:true, emit: reads
+    tuple val(meta), path('*.json')           , emit: json
```
Comment (taylorreiter) [inline modules/nf-core/fastp/main.nf:17]: just curious -- what do the `emit`s do?

File: modules/nf-core/metabat2/jgisummarizebamcontigdepths/main.nf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    path "versions.yml"                    , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    export OMP_NUM_THREADS=$task.cpus
```
Comment (taylorreiter) [inline modules/nf-core/metabat2/jgisummarizebamcontigdepths/main.nf:24]: again i know this is probs fine because its on nf-core but what does this do? I don't see where it's called or used?

File: modules/nf-core/spades/main.nf [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    fi
+    if [ -f assembly_graph_with_scaffolds.gfa ]; then
+        mv assembly_graph_with_scaffolds.gfa ${prefix}.assembly.gfa
+        gzip -n ${prefix}.assembly.gfa
+    fi
+
+    if [ -f gene_clusters.fasta ]; then
+        mv gene_clusters.fasta ${prefix}.gene_clusters.fa
+        gzip -n ${prefix}.gene_clusters.fa
+    fi
```
Comment (taylorreiter) [inline modules/nf-core/spades/main.nf:66]: oh I HATE this. it's so brittle! if any of these fail, then the whole process will fail! 

i mean obviously it's fine. but i hate it.

File: nextflow.config [hunk 7]
```diff
@@ -20,6 +21,14 @@ params {
     max_multiqc_email_size     = '25.MB'
     multiqc_methods_description = null
 
+// Fastp parameters
+    fastp_save_trimmed_fail                 = false
+    fastp_qualified_quality                 = 15
+    fastp_cut_mean_quality                  = 15
+    adapterremoval_minquality               = 2
+    keep_phix                               = false
+    adapterremoval_trim_quality_stretch     = false
```
Comment (taylorreiter) [inline nextflow.config:30]: wait why is this false? what does that do?

File: nextflow.config [hunk 8]
```diff
@@ -20,6 +21,14 @@ params {
     max_multiqc_email_size     = '25.MB'
     multiqc_methods_description = null
 
+// Fastp parameters
+    fastp_save_trimmed_fail                 = false
+    fastp_qualified_quality                 = 15
```
Comment (taylorreiter) [inline nextflow.config:26]: opinion -- trimming this stringently will remove true biological diversity in the sample for low coverage organisms. I usually use 2 or 4. 2 is still a 50% chance it's accurate.

I think the only time i think of trimming more stringently is
1. variant calling
2. when I REALLY want to prove something isn't an artifact (like when i'm demonstrating my new method actually does what I say it does)

**Review by taylorreiter — APPROVED** [review]
LGTM! Some questions for my own curiousity and some opinions :) Since you have CI set up on PRs, it would be super helpful if you could tackle removing the things you expect to fail, or at least documenting what they are.

File: modules/local/bowtie2_assembly_build.nf (hunk 3 above)
Comment (elizabethmcd) [reply modules/local/bowtie2_assembly_build.nf:21]: it is used in the`mapping_depth.nf` subworkflow, which takes in the assembly first and builds the index, and then maps with the `align` module, and sorts with samtools, and then I pass to the metabat2/jgisummarizecontigs to get the depth table. It's not super obvious because in the `main.nf` workflow it called the `mapping_depth.nf` subworkflow, which is where the above is used.

File: nextflow.config (hunk 8 above)
Comment (elizabethmcd) [reply nextflow.config:26]: Hmm ok, I will change the default behavior of this and add to the parameters help menu that this can be changed to more stringent

File: nextflow.config (hunk 7 above)
Comment (elizabethmcd) [reply nextflow.config:30]: I need to check this

File: modules/nf-core/fastp/main.nf (hunk 4 above)
Comment (elizabethmcd) [reply modules/nf-core/fastp/main.nf:17]: I could be wrong but I think it is equivalent to defining an output channel - and this is used pretty consistently in nf-core modules/subworkflows, I think it's cleaner?

File: modules/nf-core/metabat2/jgisummarizebamcontigdepths/main.nf (hunk 5 above)
Comment (elizabethmcd) [reply modules/nf-core/metabat2/jgisummarizebamcontigdepths/main.nf:24]: In the subworkflow `mapping_depth.nf` - it feels hidden, maybe I should add some comments in the `main.nf` workflow at the mapping and depth step what is actually being done in this subworkflow

File: CITATIONS.md (hunk 1 above)
Comment (elizabethmcd) [reply CITATIONS.md:24]: I went to the samtools github and they list that publication for citing: https://github.com/samtools/samtools#citing so I went with that 🤷‍♀️

File: modules/local/bowtie2_assembly_align.nf (hunk 2 above)
Comment (mertcelebi) [reply modules/local/bowtie2_assembly_align.nf:1]: I'd add this comment to the file itself and point out where it differs from the nf-core version

File: subworkflows/local/mapping_depth.nf [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+workflow MAPPING_DEPTH {
+    take:
+    assemblies
+    reads
+
+    main:
+    ch_versions = Channel.empty()
+
+    // build index
+    BOWTIE2_ASSEMBLY_BUILD ( assemblies )
```
Comment (mertcelebi) [inline subworkflows/local/mapping_depth.nf:19]: What is the convention for the spacing around these calls? They're kinda all over the place

File: subworkflows/local/mapping_depth.nf [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                def meta_new = meta.clone()
+                meta_new['binner'] = 'MetaBAT2'
+                [meta_new, depths]
+                }
+
+    // emit results
+    emit:
+    grouped_mappings                = ch_grouped_mappings
+    metabat_depths                  = ch_metabat_depths
+
```
Comment (mertcelebi) [inline subworkflows/local/mapping_depth.nf:56]: Unnecessary spaces

**Review by mertcelebi — COMMENTED** [review]
This looks generally fine, but I have some philosophical questions:
* A lot of the nf-core modules etc come with support for singularity etc etc. Do we have to support these? Can we just clean those up and remove from both the config files and processes?
* Are these even meaningful given our portions don't support these? 
* I am a little weary of having seeming support for all sorts of "profiles" without having the need and bandwidth to make sure that all of them works. Something even as simple as the gitpod profile etc etc.
* The most likely support for all our pipelines should be Docker, conda/mamba and AWS Batch. Perhaps local as well. But that's about it I think.

**Comment by elizabethmcd** [comment]
The large issue with not including profiles for Singularity is that many university/government HPC systems do not allow Docker and only Singularity. On those systems you can use singularity tools to convert docker images to singularity, but that would require a bit of work on the users end if they are trying to use our workflows. I don't know how to reconcile providing those profiles in our workflows without testing them. I'm not sure how nf-core handles this if they are going through and testing that each profile works or if this is handled at the module level where you declare the image/environment used.

File: subworkflows/local/mapping_depth.nf (hunk 9 above)
Comment (elizabethmcd) [reply subworkflows/local/mapping_depth.nf:19]: Scattered across nf-core pipelines and modules (and I assume with subworkflows to come) is the non-consistent use of spacing, or extra spacing. Currently I assume they do this for readability but it actually makes things cumbersome down the line so I will get rid of extra spacing.

**Comment by taylorreiter** [comment]
My interpretation of the profiles is that especially if we're mostly building pipelines from nf-core modules, it's a light lift to maintain all of the profiles. Is this wrong @elizabethmcd?

**Comment by elizabethmcd** [comment]
That's my understanding is that building from nf-core modules and implementing their practices it's a light lift to maintain those profiles. The question becomes if we need to make a new module do we first go through the nf-core modules repo and PR process before then implementing that in one of our workflows. I think this becomes relevant with the new `--with-wave` functionality that you can build, deploy, and test the container when launching the workflow so in that case we would only be testing Docker images and the docker profile.

**Comment by mertcelebi** [comment]
>The question becomes if we need to make a new module do we first go through the nf-core modules repo and PR process before then implementing that in one of our workflows.

What does this mean? Like do you mean if we should first publish our submodules to `nf-core` before incorporating to our own? If so, that would be crazy. That seems unnecessarily convoluted unless we are doing something with the explicit goal of contributing to the `nf-core` repository. 

> it's a light lift to maintain all of the profiles

I don't think this is true. Especially if we end up building bunch of new modules as part of our pipelines (most likely the case). For pipelines that we pull from `nf-core`, we don't have a lot of visibility into how the containers `nf-core` provides works. For Docker containers, we'll most likely catch any issues, because we'll be actively using these. But I worry that we won't catch issues with un-used parts of the profiles (since there are so many!) and the added profiles will only create friction and tech-debt.

**Comment by elizabethmcd** [comment]
1. Yes an option would be to first publish modules/subworkflows to `nf-core` before incorporating into our workflows, and yes I agree this is convoluted. I think if we have a new module we develop it internally and maybe once stable and could be useful for the community it goes through the process of being contributed to `nf-core`. This way we avoid lag times with PRs through `nf-core` 
2. Maintaining profiles, there are two cases: 
    - If a workflow uses modules all available with `nf-core` should we keep those extra profiles? Or rip them out regardless? 
    - If we have a workflow with new modules for which we need new containers, I agree we should just start with the Docker profile and then work upwards if needed. However this brings back the issue that on many university/gov HPC systems Docker won't work and we might have to add and test Singularity functionality. The others (like podman etc.) I'm fine with removing
  
Another point to consider - if we create containers should we be contributing these back to repos like BioContainers? My vote is yes but this requires some extra work and guidance.
